### PR TITLE
Google Cloud Artifact Registry support for beetmoverscript

### PR DESCRIPTION
**Pull request by gabrielbusta**
### Description

This is what a `task.json` looks like:
```
{
  ...
  "scopes": [
    "project:releng:beetmover:apt-repo:nightly",
    "project:releng:beetmover:action:import-from-gcs-to-artifact-registry"
  ],
  "payload": {
    "product": "firefox",
    "gcs_sources": [
      "pub/firefox/nightly/2023/02/2023-02-23-18-33-31/firefox-112.0a1.en-US.linux-x86_64.deb"
    ],
    ...
  },
  ...
}
```
### Reference

- [Bug 1799763](https://bugzilla.mozilla.org/show_bug.cgi?id=1799763)
- [Bug 1799769](https://bugzilla.mozilla.org/show_bug.cgi?id=1799769)
- [RELENG-1042](https://mozilla-hub.atlassian.net/browse/RELENG-1042)
- [SVCSE-1067](https://mozilla-hub.atlassian.net/browse/SVCSE-1067)
- [gecko rev. D172619](https://phabricator.services.mozilla.com/D172619)

### CODE REVIEW

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 1]
```diff
@@ -101,6 +118,72 @@ async def upload_to_gcs(context, target_path, path):
...
+        return context.gcs_client.import_apt_artifacts(request=import_artifacts_request)
+    if context.resource_type == "yum-repo":
+        return context.gcs_client.import_yum_artifacts(request=import_artifacts_request)
+    raise Exception("Artifact Registry resource must be one of [apt-repo, yum-repo]")
+
+
+async def import_artifacts(context):
+    """Imports repo artifacts from cloud storage to artifact registry"""
+    product = get_product_name(context.task, context.config)
+    project = "moz-fx-dev-releng"  # the project the repo is in. hard-coding our dev. env. for now
```
Comment (gabrielBusta) [inline beetmoverscript/src/beetmoverscript/gcloud.py:160]: ah I forgot about this, this needs some work

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 2]
```diff
@@ -101,6 +118,72 @@ async def upload_to_gcs(context, target_path, path):
...
+    if context.resource_type == "apt-repo":
+        return context.gcs_client.import_apt_artifacts(request=import_artifacts_request)
+    if context.resource_type == "yum-repo":
+        return context.gcs_client.import_yum_artifacts(request=import_artifacts_request)
+    raise Exception("Artifact Registry resource must be one of [apt-repo, yum-repo]")
+
+
+async def import_artifacts(context):
+    """Imports repo artifacts from cloud storage to artifact registry"""
+    product = get_product_name(context.task, context.config)
```
Comment (gabrielBusta) [inline beetmoverscript/src/beetmoverscript/gcloud.py:159]: Ah I forgot I hard-coded the GCP project and location, this needs some work

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 3]
```diff
@@ -101,6 +118,72 @@ async def upload_to_gcs(context, target_path, path):
...
+    project = "moz-fx-dev-releng"  # the project the repo is in. hard-coding our dev. env. for now
+    location = "northamerica-northeast2"  # this is the region / location of the repo. got it from the console
```
Comment (gabrielbusta) [inline beetmoverscript/src/beetmoverscript/gcloud.py]: Ah I forgot I hard-coded the GCP `project` and `location`, this needs some work

File: beetmoverscript/docker.d/init_worker.sh [hunk 4]
```diff
@@ -21,6 +21,7 @@ case $COT_PRODUCT in
         test_var_set 'DEP_PARTNER_KEY'
         test_var_set 'MAVEN_ID'
         test_var_set 'MAVEN_KEY'
+        test_var_set 'PROJECT_NAME'
```
Comment (gabrielbusta) [inline beetmoverscript/docker.d/init_worker.sh]: ~~more details about this env var [here](https://mozilla-hub.atlassian.net/browse/SVCSE-1067)~~

Comment (gabrielbusta) [reply beetmoverscript/docker.d/init_worker.sh]: This is not what I thought... looked in the `configmap` and the value is `beetmover`

File: beetmoverscript/requirements/base.in [hunk 5]
```diff
@@ -5,4 +5,5 @@ mozilla-version
 redo
 scriptworker
 taskcluster
-google-cloud-storage
\ No newline at end of file
+google-cloud-storage
+google-cloud-artifact-registry
```
Comment (hneiva) [inline beetmoverscript/requirements/base.in]:
```suggestion
google-cloud-artifact-registry

```
Missing empty line at EOF

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 6]
```diff
@@ -72,7 +76,21 @@ def handle_exception(e):
...
+    client = artifactregistry_v1.ArtifactRegistryAsyncClient()
+    return client
+
+
+def set_gcp_client(context):
+    product = get_product_name(context.task, context.config)
+    if "repo" in context.resource_type:
+        context.gcp_client = _get_artifact_registry_client(context, product)
+    else:
+        context.gcp_client = _get_gcs_client(context, product)
```
Comment (hneiva) [inline beetmoverscript/src/beetmoverscript/gcloud.py]: `gcp_client` has very different methods when dealing with the different services.  `context` is bad enough as a god object, and we made this mistake with "bucket name" property already.

I suggest we split it into `gcs_client` and `gar_client` (or something of the sorts).

```suggestion
        context.gar_client = _get_artifact_registry_client(context, product)
        ....
        context.gcs_client = _get_gcs_client(context, product)
```

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 7]
```diff
@@ -72,7 +76,21 @@ def handle_exception(e):
...
+
+def _get_artifact_registry_client(context, product):
+    """Set up a google-cloud-artifact-registry client"""
+    client = artifactregistry_v1.ArtifactRegistryAsyncClient()
+    return client
+
+
+def set_gcp_client(context):
+    product = get_product_name(context.task, context.config)
+    if "repo" in context.resource_type:
```
Comment (hneiva) [inline beetmoverscript/src/beetmoverscript/gcloud.py]:
```suggestion
    if context.resource_type in ("yum-repo", "apt-repo"):
```

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 8]
```diff
@@ -101,12 +119,74 @@ async def upload_to_gcs(context, target_path, path):
...
+def build_artifact_registry_import_artifacts_request(context, repository, gcs_source):
+    import_request_kwargs = {
+        "gcs_source": gcs_source,
+        "parent": repository.name,
+    }
+    if context.resource_type == "apt-repo":
+        return artifactregistry_v1.ImportAptArtifactsRequest(**import_request_kwargs)
+    if context.resource_type == "yum-repo":
+        return artifactregistry_v1.ImportYumArtifactsRequest(**import_request_kwargs)
+    # artifact registry supports docker... interesting...
```
Comment (hneiva) [inline beetmoverscript/src/beetmoverscript/gcloud.py]:
```suggestion
```

File: beetmoverscript/src/beetmoverscript/task.py [hunk 9]
```diff
@@ -64,34 +67,34 @@ def _check_scopes_exist_and_all_have_the_same_prefix(scopes, prefixes):
...
 
 
-def get_task_bucket(task, script_config):
-    """Extract task bucket from scopes"""
-    prefixes = _get_bucket_prefixes(script_config)
-    scopes = _extract_scopes_from_unique_prefix(task["scopes"], prefixes=prefixes)
-    buckets = [s.split(":")[-1] for s in scopes for p in prefixes if s.startswith(p)]
-    log.info("Buckets: %s", buckets)
+def get_task_resource(context):
+    """Extract task cloud resource from scopes"""
```
Comment (hneiva) [inline beetmoverscript/src/beetmoverscript/task.py]: Please add an explanation of what "resource" is in this context.

Also: Thank you for renaming this. It was VERY confusing.

File: beetmoverscript/src/beetmoverscript/utils.py [hunk 10]
```diff
@@ -155,7 +160,7 @@ def get_product_name(task, config, lowercase_app_name=True):
 
     action = get_task_action(task, config)
 
-    if action == "push-to-releases":
+    if action == "push-to-releases" or action == "import-from-gcs-to-artifact-registry":
```
Comment (hneiva) [inline beetmoverscript/src/beetmoverscript/utils.py]:
```suggestion
    if action in ("push-to-releases", "import-from-gcs-to-artifact-registry"):
```

File: beetmoverscript/src/beetmoverscript/gcloud.py [hunk 11]
```diff
@@ -101,12 +119,74 @@ async def upload_to_gcs(context, target_path, path):
...
+    repository = await context.gcp_client.get_repository(request=get_repo_request)
+    log.info(repository)
+
+    gcs_source = build_artifact_registry_gcs_source(context, product)
+    log.info(gcs_source)
+
+    import_artifacts_request = build_artifact_registry_import_artifacts_request(context, repository, gcs_source)
+    log.info(import_artifacts_request)
+
+    async_operation = await do_artifact_registry_import_artifacts_request(context, import_artifacts_request)
```
Comment (hneiva) [inline beetmoverscript/src/beetmoverscript/gcloud.py]: Looking at `build_artifact_registry_import_artifacts_request`, `build_artifact_registry_gcs_source` and `do_artifact_registry_import_artifacts_request`, makes me wish we were more object-oriented (with classes).
All 3 functions are switching on the same flag, and "doing more than one thing".

One way of simplifying  the logic would be to have the switching in `import_from_gcs_to_artifact_registry`
```
if context.resource_type == "apt-repo":
    result = import_apt(....)
elif context.resource_type == "yum-repo":
    result = import_yum(...)
else:
    raise Exception("Artifact Registry resource must be one of [apt-repo, yum-repo]")
```

**Review by hneiva — CHANGES_REQUESTED** [review]

**Comment by gabrielbusta** [comment]
[Here is a try push beetmovin' the langpacks and the binary.](https://treeherder.mozilla.org/jobs?repo=try&revision=5d522f7c362cf62abcb83e052df80f78892c1566)

These are `beetmover-apt` tasks [1](https://firefox-ci-tc.services.mozilla.com/tasks/VsNJ2I1DRYGg0OfTuyJQag), [2](https://firefox-ci-tc.services.mozilla.com/tasks/elRI4ehrSimfpabymm88AQ), and [3](https://firefox-ci-tc.services.mozilla.com/tasks/UQ93sT7YS_SN360FFhuhyA).

[Here is a staging beta beetmovin' the langpacks and the binary.](https://treeherder.mozilla.org/jobs?repo=try&revision=a72e7b08483db074c63f0d8c47730755465b0d66)

This is the [beta's beetmover-apt task](https://firefox-ci-tc.services.mozilla.com/tasks/LeYgYtgzQ-ih8SwGqt_Faw) (the staging release only generates a subset of the langpacks)

**Review by hneiva — APPROVED** [review]

**Pull request merged**